sdk/java-sdk/src/main/java/kalix/javasdk/workflow/Workflow.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright 2021 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kalix.javasdk.workflow;
+
+import akka.annotation.ApiMayChange;
+import io.grpc.Status;
+import kalix.javasdk.DeferredCall;
+import kalix.javasdk.Metadata;
+import kalix.javasdk.impl.workflow.WorkflowEffectImpl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+public abstract class Workflow<S> {
+
+
+  private Optional<CommandContext> commandContext = Optional.empty();
+
+  private Optional<S> currentState = Optional.empty();
+
+  private boolean handlingCommands = false;
```

Comment:
do we need it?

Reply:
Yes and no. 

It's wrong to call it `handlingCommands` but we do need such a flag. Accessing the `currentState` should only be possible when running a step or handling a command. Calling this method in the constructor, for instance, is illegal. 

The state is only available once we have a message to deliver to the underlying component.